models/layers.py

```diff
@@ -0,0 +1,123 @@
+import torch
+import numpy as np
+from torch import nn
+from torch.autograd import Variable
+
+# from torch.nn import functional as F
+
+
+class CNN_Encoder(nn.Module):
+    def __init__(self, output_size, input_size=(1, 28, 28)):
+        super(CNN_Encoder, self).__init__()
+
+        self.input_size = input_size
+        self.channel_mult = 16
+
+        if self.input_size[1] == 32:
+            self.conv = nn.Sequential(
+                nn.Conv2d(in_channels=3, out_channels=self.channel_mult * 1, kernel_size=4, stride=1, padding=1),
+                nn.LeakyReLU(0.2, inplace=True),
+                nn.Conv2d(self.channel_mult * 1, self.channel_mult * 2, 4, 2, 1),
+                nn.BatchNorm2d(self.channel_mult * 2),
+                nn.LeakyReLU(0.2, inplace=True),
+                nn.Conv2d(self.channel_mult * 2, self.channel_mult * 4, 4, 2, 1),
+                nn.BatchNorm2d(self.channel_mult * 4),
+                nn.LeakyReLU(0.2, inplace=True),
+                nn.Conv2d(self.channel_mult * 4, self.channel_mult * 8, 4, 2, 1),
+                nn.BatchNorm2d(self.channel_mult * 8),
+                nn.LeakyReLU(0.2, inplace=True),
+            )
+        else:
+            self.conv = nn.Sequential(
+                nn.Conv2d(in_channels=1, out_channels=self.channel_mult * 1, kernel_size=4, stride=1, padding=1),
+                nn.LeakyReLU(0.2, inplace=True),
+                nn.Conv2d(self.channel_mult * 1, self.channel_mult * 2, 4, 2, 1),
+                nn.BatchNorm2d(self.channel_mult * 2),
+                nn.LeakyReLU(0.2, inplace=True),
+                nn.Conv2d(self.channel_mult * 2, self.channel_mult * 4, 4, 2, 1),
+                nn.BatchNorm2d(self.channel_mult * 4),
+                nn.LeakyReLU(0.2, inplace=True),
+                nn.Conv2d(self.channel_mult * 4, self.channel_mult * 8, 4, 2, 1),
+                nn.BatchNorm2d(self.channel_mult * 8),
+                nn.LeakyReLU(0.2, inplace=True),
+                nn.Conv2d(self.channel_mult * 8, self.channel_mult * 16, 3, 2, 1),
+                nn.BatchNorm2d(self.channel_mult * 16),
+                nn.LeakyReLU(0.2, inplace=True),
+            )
+
+        self.flat_fts = self.get_flat_fts(self.conv)
```

Comment:
What's "fts"? :)

Reply:
I believe 'fts' refers to features. It calculates the size of flattened layers. It's a function I left from the initial repo and didn't give much thought towards the naming

Reply:
That makes sense. Do you know why is this value different from the `output_size`? Looks like it should be equal to the `embeddings_size`, which I would guess to be equal to the encoder output size :thinking:

Reply:
`output_size` is the final representation created by the encoder. So, while `fts` represents the number of features from the convolutional part we still have an additional linear layer to go through
```{Python}
self.linear = nn.Sequential(
            nn.Linear(self.flat_fts, output_size),
            nn.BatchNorm1d(output_size),
            nn.LeakyReLU(0.2),
        )
```
At least that's how I understand it